File: python/src/base_human/basen.py
"""Fixed-length base-N encode and decode, spec sections 5.1 through 5.3."""

from __future__ import annotations

from .errors import INVALID_CHARACTER, OUT_OF_RANGE, HrcError


def alphabet_index(alphabet: str) -> dict:
    return {ch: i for i, ch in enumerate(alphabet)}


def encode_base_n(value: int, alphabet: str, length: int) -> str:
    """Fixed-length base-N encode, most significant digit first."""
    base = len(alphabet)
    capacity = base ** length
    if value < 0 or value >= capacity:
        raise HrcError(OUT_OF_RANGE, "value is outside the fixed-length capacity")
    out = [""] * length
    v = value
    for pos in range(length - 1, -1, -1):
        out[pos] = alphabet[v % base]
        v //= base
    return "".join(out)


def decode_base_n(text: str, alphabet: str, index: dict | None = None) -> int:
    base = len(alphabet)
    if index is None:
        index = alphabet_index(alphabet)
    value = 0
    for ch in text:
        digit = index.get(ch)
        if digit is None:
            raise HrcError(INVALID_CHARACTER, f"Symbol {ch!r} is not in the alphabet")
        value = value * base + digit
    return value

File: ruby/lib/baseh/basen.rb
# frozen_string_literal: true

module Baseh
  # Fixed-length base-N encoding, spec section 5. Most significant digit first.
  module BaseN
    module_function

    # Spec 5.1. All arithmetic stays in Integer (arbitrary precision).
    def encode_base_n(value, alphabet, length)
      base = alphabet.length
      out = Array.new(length)
      v = value
      (length - 1).downto(0) do |pos|
        digit = v % base
        out[pos] = alphabet[digit]
        v /= base
      end
      out.join
    end

    # Spec 5.2. Raises INVALID_CHARACTER for symbols outside the alphabet.
    def decode_base_n(text, alphabet, index = nil)
      index ||= alphabet_index(alphabet)
      base = alphabet.length
      value = 0
      text.each_char do |ch|
        digit = index[ch]
        if digit.nil?
          raise BasehError.new(
            "INVALID_CHARACTER",
            "Symbol #{ch.inspect} is not in the alphabet"
          )
        end
        value = value * base + digit
      end
      value
    end

    def alphabet_index(alphabet)
      alphabet.each_char.each_with_index.to_h
    end
  end
end
